Wires refactor 2

MMaiero opened the pull request:
Few more minor changes to the wires.
Change descriptions in respective comments.

MMaiero committed:
Small changes in wire.component.provider.

This solves some encountered issues with BLOBs and with reconcile table/column.
Changes to BaseAsset.
Removed dependency in web2.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Solved slow creation of wire graph.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

amitjoy commented on kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/asset/provider/BaseAsset.java:
```diff
@@ -321,28 +341,31 @@ public ComponentConfiguration getConfiguration() throws KuraException {
     }
 
+    @SuppressWarnings("unchecked")
+    private void fillDriverSpecificChannelConfiguration(final Tocd mainOcd,
```
Add invariants

amitjoy commented on kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/asset/provider/BaseAsset.java:
```diff
@@ -555,9 +580,14 @@ public void onDriverEvent(final DriverEvent event) {
-                    + CHANNEL_PROPERTY_POSTFIX.value();
-            channelPrefixes.add(prefix);
+
+            final StringBuilder channelPrefix = new StringBuilder();
```
Not needed. Java 8 internally uses `StringBuilder` for string concatenation.

MMaiero replied:
Not really. The string concatenation, if performed, is done at different steps (at compile time or using JIT). The current Java8 for ARM is provided without JIT, so in our target hardware, it is fully interpreted and the concatenation optimizations are not performed.
Finally, having a look [at this research](http://www.pellegrino.link/2015/08/22/string-concatenation-with-java-8.html), using explicitly StringBuilder is still the best option.

amitjoy replied:
Good source of knowledge but due to static concatenation, StringBuilder and String concatenation works equally. For dynamic concatenation, I completely agree that StringBuilder performs better than String concatenation.

amitjoy commented on kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/internal/asset/provider/AssetOptions.java:
```diff
@@ -103,11 +103,8 @@ private void addChannel(final Channel channel) {
-     *             if the argument is null
      */
     private void checkChannelAvailability(final Map<String, Object> properties) {
-        requireNonNull(properties, message.propertiesNonNull());
```
The contract of this method is to use a map which is not null. Even though the map to be passed to this method in this class is not null, the contract should apply such preconditions.

MMaiero replied:
Those methods are private. The check is performed at the boundaries and we are assured that the map is not null. I completely agree in having contracts specified for public methods, since those define the interaction, and for these it's meaningful to check what is passed.
For private methods in a class it's a little bit an overkill, with an unneeded cost in terms of CPU time.

https://dev.eclipse.org/sonar/component/index#component=org.eclipse.kura%3Aorg.eclipse.kura.asset.provider%3Asrc%2Fmain%2Fjava%2Forg%2Feclipse%2Fkura%2Finternal%2Fasset%2Fprovider%2FAssetOptions.java&currentIssue=05c789f4-1d0e-4c9c-96dd-ebfa2bdcb17a

amitjoy replied:
SonarQube is just awesome 😉

amitjoy commented on kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/internal/asset/provider/AssetOptions.java:
```diff
@@ -120,11 +117,8 @@ private void checkChannelAvailability(final Map<String, Object> properties) {
-     *             if the argument is null
      */
     private void extractProperties(final Map<String, Object> properties) {
-        requireNonNull(properties, message.propertiesNonNull());
```
Same here

amitjoy commented on kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/internal/asset/provider/AssetOptions.java:
```diff
@@ -158,9 +152,10 @@ public AssetConfiguration getAssetConfiguration() {
-        requireNonNull(properties, message.propertiesNonNull());
-        requireNonNull(channelTypePropertyKey, message.channelKeyNonNull());
+    private ChannelType getChannelType(final Map<String, Object> properties, final String channelKeyFormat) {
+        requireNonNull(channelKeyFormat, message.channelKeyNonNull());
```
Same

amitjoy commented on kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/internal/asset/provider/AssetOptions.java:
```diff
@@ -243,50 +237,56 @@ private Channel retrieveChannel(final long channelId, final Map<String, Object>
         if (channelId <= 0) {
             throw new IllegalArgumentException(message.channelIdNotLessThanZero());
         }
-        requireNonNull(properties, message.propertiesNonNull());
```
This precondition is feasible here

amitjoy commented on kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/internal/asset/provider/AssetOptions.java:
```diff
@@ -243,50 +237,56 @@ private Channel retrieveChannel(final long channelId, final Map<String, Object>
-        final String channelKeyContainment = CHANNEL_PROPERTY_POSTFIX.value() + CHANNEL_PROPERTY_PREFIX.value()
-                + CHANNEL_PROPERTY_POSTFIX.value();
-        final String channelKeyFormat = channelId + channelKeyContainment;
+        final StringBuilder channelPropertyKeyBuilder = new StringBuilder();
```
Java 8 uses `StringBuilder` by default

amitjoy commented on kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/internal/asset/provider/AssetOptions.java:
```diff
@@ -243,50 +237,56 @@ private Channel retrieveChannel(final long channelId, final Map<String, Object>
         return channel;
     }
 
+    private Map<String, Object> retrieveChannelConfig(final long channelId, final Map<String, Object> properties) {
```
Add preconditions

amitjoy commented on kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/internal/asset/provider/AssetOptions.java:
```diff
@@ -243,50 +237,56 @@ private Channel retrieveChannel(final long channelId, final Map<String, Object>
+        return channelConfig;
+    }
+
+    private String retrieveChannelName(final Map<String, Object> properties, final String channelKeyFormat) {
```
Add preconditions

amitjoy commented on kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/client/ui/wires/PropertiesUi.java:
```diff
@@ -628,10 +628,8 @@ private int findNextChannelId() {
             public String getValue(final GwtChannelInfo object) {
-                final String value = object.get(id).toString();
-                return value == null ? "" : value;
+                return object.get(id).toString();
```
`object.get(id)` can return null and result in NPE

MMaiero replied:
if I try to execute this `final String value = object.get(id).toString();` and `object.get(id)` is null I would end in a NPE in any case when calling the `toString`.
Having `return value == null ? "" : value;` the following line would not prevent it.

amitjoy replied:
Exactly. Previously also, it was erroneous. Just add a Null Check If condition on the get to avoid it.

amitjoy commented on kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/common/DbServiceHelper.java:
```diff
@@ -76,7 +76,6 @@ public static DbServiceHelper getInstance(final DbService dbService) {
      *             if argument is null
      */
     public void close(final Connection conn) {
-        requireNonNull(conn, message.connectionNonNull());
```
Shouldn't it be a precondition to this method? Even though `DbService` `close()` method doesn't work on null elements, my proposition would be too add invariant to this method. I believe this is a good practice.

MMaiero replied:
The check if Connection is null is already performed in the Db implementation. So, having the check here is not needed. Furthermore, we should try to keep consistency and this is not the case if we throw an exception when trying to close the connection, while we accept to close a null statement.

amitjoy replied:
According to my opinion, accepting a null is a violation of contract and it does not symbolize the nature of true object orientation. That's why such invariant is very helpful in implementing methods. I agree that `DbService.close()` method also accepts null but does not work on it because of its null check but I personally don't feel it's a good practice when it's a public API.

MMaiero commented:
I've tried to do some more steps. Please have a look.
I think that the commented aspects are, in some way, minor elements that would not affect the overall correctness or performance. I believe that, if most of the elements are addressed, I would like to merge and, eventually, solve later on. In this way, we can speed up the M1 release.

amitjoy commented on kura/org.eclipse.kura.asset.cloudlet.provider/src/main/java/org/eclipse/kura/internal/asset/cloudlet/AssetCloudlet.java:
```diff
@@ -427,27 +430,29 @@ private void wrapValue(final AssetRecord assetRecord, final String userValue, fi
         requireNonNull(userValue, message.valueNonNull());
         requireNonNull(userType, message.typeNonNull());
 
+        final DataType dataType = DataType.getDataType(userType);
```
Add this inside the `try` block as it might throw `IllegalArgumentException` and the exception must be propagated upwards. Also catch `IllegalArgumentException` in this method as well as in `doPutAssets()` in the cloudlet.

amitjoy commented on kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/asset/provider/BaseAsset.java:
```diff
@@ -320,27 +339,41 @@ public ComponentConfiguration getConfiguration() throws KuraException {
+    @SuppressWarnings("unchecked")
+    private void fillDriverSpecificChannelConfiguration(final Tocd mainOcd,
+            final List<Tad> driverSpecificChannelConfiguration) {
+        if (mainOcd == null || driverSpecificChannelConfiguration == null) {
```
`mainOcd == null || driverSpecificChannelConfiguration == null || driverSpecificChannelConfiguration.isEmpty()`

amitjoy replied:
This `isEmpty` check reduces the unnecessary variable references creation that happens after this.

MMaiero replied:
I think that the code manages the case when `driverSpecificChannelConfiguration` is empty. 
Furthermore, having again a look at the code, I think that in another refactoring pass we can improve the methods structuring and have them more specific and more understandable. 
But this will go for M2. M1 is mostly closed now.

amitjoy commented on kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/internal/asset/provider/AssetOptions.java:
```diff
@@ -188,39 +170,17 @@ private ChannelType getChannelType(final Map<String, Object> properties, final S
      *             if any of the arguments is null
      */
-    private DataType getDataType(final Map<String, Object> properties, final String channelValueTypePropertyKey) {
+    private DataType getDataType(final Map<String, Object> properties, final String channelKeyFormat) {
```
`@throws IllegalArgumentException if the Data Type is not found`

MMaiero committed:
Fixed possible memory leak related to not closed result sets.

Small modification to the wire filter in order to not cache empty WireRecords.
Created utility method in ChannelType and DataType.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

amitjoy commented on kura/org.eclipse.kura.asset.cloudlet.provider/src/main/java/org/eclipse/kura/internal/asset/cloudlet/AssetCloudlet.java:
```diff
@@ -292,9 +295,9 @@ private void doPutAssets(final KuraRequestPayload reqPayload, final KuraResponse
             try {
                 this.wrapValue(assetRecord, userValue, userType);
-            } catch (final NumberFormatException nfe) {
+            } catch (final IllegalArgumentException e) {
```
It should be `catch (final IllegalArgumentException | NumberFormatException  e)`

MMaiero replied:
Please have a look at the javadoc https://docs.oracle.com/javase/7/docs/api/java/lang/NumberFormatException.html

amitjoy replied:
👍

amitjoy commented:
ok cool. I will have a final look now and will merge by today if there is no further significant comment.
